### ISSUE TITLE
Variables @SUBJECT@ and others - decoding problems

### ISSUE THREAD

**Issue by kodorevi8ulis** [RealRaven2000/FiltaQuilla#102]
Is there any way to use regexp for variables like @subject@ ?

for example i use an app to notify me through telegram from specific emails 
example -> c:\notify.exe,@SUBJECT@,CustomText

the problem is that sometimes the subject contains non latin chars which aren't passed correctly 
and also , i want to get only a part of the subject using regular expressions

example -> the subject contains various tags which i want to exclude 
like ->  [Tag1] [Tag2] - Text1 @ Text2 
and i want only the Text1 and Text2

**Comment by RealRaven2000** [comment]
> Is there any way to use regexp for variables like @subject@ ?
> 

no there isn't.

> for example i use an app to notify me through telegram from specific emails
> example -> c:\notify.exe,@subject@,CustomText
> 

what does it do?

> the problem is that sometimes the subject contains non latin chars which aren't passed correctly
> and also , i want to get only a part of the subject using regular expressions
> 

sounds complicated an specialized. I donp't think many users would benefit from it. But we can look into solving the non-latin characters problem first, maybe there is a decoding issue. Can you give an example for that?

**Comment by kodorevi8ulis** [comment]
the app just notifies me through a telegram bot for some emails

for example i get emails like this 
[Tag1] [2020-01-01] [Tag2] Μια ωραία πεταλούδα email @Store
which contains greek chars and i can't read them 
in this example i want to get something like this Store - Μια ωραία πεταλούδα email

so i use the app to send me an message with title Store and text Μια ωραία πεταλούδα email

now i actually do it with a "weird" and no so fast way
i save the email as a file , then i use the app to get the subject (usually a base64 string so i also decode it.... maybe it's easier to create an utf @subject64@ var? ) , send me the message , then delete the eml etc

**Comment by Null-Kelvin** [comment]
Hi, I have similar problem with non-latin characters in @AUTHOR@ and @SUBJECT@ variables. For example, when I get email with subject "Дистрибы" in my vbs-script I have string like "1@>A8<>2 5>=84 "
Or, subject "Другие способы познакомиться с Wrike" converted to "@C385 A?>A>1K ?>7=0:><8BLAO A Wrike".

**Comment by DebuggedTBonce** [comment]
The world is getting smaller and smaller... and **non-latin** subjects (and even addresses!) are **extremely common** these days: I think RealRaven2000 is wrong believing that fixing non-latin in @SUBJECT@ etc. would "not benefit many users". It definitely would: think of Europe (bigger than the US) with their "weird" languages, not to mention China, India, etc.

My own equivalent of kodorevi8ulis's "c:\notify.exe" _can read Unicode_ parameters - but alas, FiltaQuilla sends them already truncated (in a way I don't understand, btw) to the executable in the Run Program filter.

Could the @[...]@ variables be sent in Unicode, perhaps in an additional "Run Program _Unicode_" filter to preserve backward compatibility?
Or else, kodorevi8ulis' idea of @subject64@ besides @subject@ is pretty good as well.

Thank you!!

**Comment by RealRaven2000** [comment]
> Hi, I have similar problem with non-latin characters in @author@ and @subject@ variables. For example, when I get email with subject "Дистрибы" in my vbs-script I have string like "�1@>A8<>2 �5>=84 " Or, subject "Другие способы познакомиться с Wrike" converted to "�@C385 A?>A>1K ?>7=0:><8BLAO A Wrike".

could you upload a test email? test data is always hard to get. Also if you have a filter definition that deals with you can export it with quickFilters or just make a screenshot.

**Comment by DebuggedTBonce** [comment]
Raven, it is actually **trivial** to create a **test mail at your end**. Here's what I've done, in Win10:
- simply copied the Russian text from your very post (i.e. your quote of of Null-Kelvin) into the clipboard - yes, it works perfectly
- created a new message in TB and pasted the Russian text into the Subject.
Here's the result, i.e. your test email:
![image](https://user-images.githubusercontent.com/92902505/138324438-62ee33c7-71d7-4892-aa9a-506255cbff97.png)

**Comment by DebuggedTBonce** [comment]
Apologies, but is this thread DEAD??

When I call my program from the debugger with "Другие способы познакомиться с Wrike"... I do collect "Другие способы познакомиться с Wrike" inside the program. But when I send myself a mail with that same subject, @SUBJECT@ arrives garbled to my program.

The question is then, **what exactly is sent in @SUBJECT@** by FiltaQuilla? It cannot be UTF16 (i.e. widechars), because that's what works fine in my program. Ergo... what is it?

Thank you!!

**Comment by RealRaven2000** [comment]
> Apologies, but is this thread DEAD??
> 
> When I call my program from the debugger with "Другие способы познакомиться с Wrike"... I do collect "Другие способы познакомиться с Wrike" inside the program. But when I send myself a mail with that same subject, @subject@ arrives garbled to my program.

no I haven't forgot about this - I am just trying to test it - as do not have a notify.exe I am trying to find a way where the subject can be meaningfully transferred and tested to a standard windows program such as "C:\Windows\system32\notepad.exe ". Unfortunately I do not know how that would react if I passed kyrillic characters as a filename which is what it's command line parameter expects... I will see what google spits out.

**Comment by RealRaven2000** [comment]
@DebuggedTBonce can I use this / is this safe for testing?
https://download.cnet.com/Notify/3000-2381_4-10324832.html

**Comment by RealRaven2000** [comment]
Just debugging this now (even if it will fail in notepad) I am using the function `hdr.mime2DecodedSubject` to retrieve the subject header, the value still looks alright within the debugger window during the call of my helper function `_replaceParameters()`:

![image](https://user-images.githubusercontent.com/6314674/139421414-31bec492-ba85-4685-8671-26984f40cca4.png)

This is being passed on unchanged in a parameters Array to the undocumented `Ci.nsIProcess,run( )` function. So far so good, but this is where the blackboxed hacking starts...

notepad isn't particularly helpful with this error message:
![image](https://user-images.githubusercontent.com/6314674/139421914-8cfb170c-bce5-459a-a7b6-214319c86077.png)

it doesn't tell me what it received as command line parameter. So I really need a (preferably free) windows app that can take any text as a parameter.

**Comment by RealRaven2000** [comment]
I also tried with wordpad.exe and winword.exe but they do not understand cyrillic as a command line argument. It turns out that my editor notepad++ has a optional cmd argument -qt="text" but I cannot make that work with the double quotes, I tried this but it throws and exception (filter failed) in Thunderbird:

C:\Program Files (x86)\Notepad++\notepad++.exe-qt=",@SUBJECT@,"

the problem here is that FiltaQuilla needs the comma separated list for arguments and do not know how to generate the double quotes through hacking the command (yet) .

**Comment by Null-Kelvin** [comment]
I tested it with "FQTest.cmd":

```
@ECHO OFF
cd /d %~dp0
chcp 65001 >NUL 2>NUL
ECHO %1 > Output.txt
ECHO %2 >> Output.txt
```
In FiltaQuilla I run that with "G:\FQTest.cmd,@AUTHOR@,@SUBJECT@"
Garbage text would be in "Output.txt"

**Comment by RealRaven2000** [comment]
I invested a few hours into the codepage issue but didn't get anywhere. Apparently there is no codepage parameter for passing to the `Ci.nsIProcess.run( )` command I am using. I also tried decoding from UTF-8 beforehand, but still just garbage on the command line:

```javascript
    function convertFromUnicode(aSrc) {  // test
      let unicodeConverter = Cc["@mozilla.org/intl/scriptableunicodeconverter"].createInstance(
        Ci.nsIScriptableUnicodeConverter
      );
      unicodeConverter.charset = "UTF-8";
      return unicodeConverter.ConvertFromUnicode(aSrc);
    };
```

**Comment by DebuggedTBonce** [comment]
[Apologies for not replying in the thread, I'm traveling and don't have my login here.]

I got it figured out! It's actually /trivial/, once you're willing to use Win's API. The "/great secret/" is to do *everything in Unicode*, using *wide chars/strings*. There are *_no codepages_ involved*. Here goes:

First and foremost, Thunderbird (TB) is nearly fully Unicode, as witnessed by the following TB shot:



- where TB correctly displays a wide variety of fonts in all sorts of languages (including emojis!). Ergo the *info is available* in TB in the proper form to send along by FiltaQuilla.

I'm using Embarcadero C++ Builder (but that's /immaterial/ to the issue, as I explain below), where I wrote a trivial application that displays program parameters received *in Unicode*. It looks like this, with the parameters specified in the IDE:



- where the "exotic" languages of the top shot are perfectly reproduced. Do note the crucial point that all the content of that window is in Unicode, using wide-char (C's *wchar_t*) variables.

Having the "exotics" in the application, I have put myself in the position of FiltaQuilla. So the next question is *how to pass those 
Unicode strings along* to a program called by FiltaQuilla. It turns out to be frankly trivial, once *wchar_t* is adopted everywhere.
Indeed, clicking the /Spawn/ button calls the following trivial function:



The function is very short indeed... but there are a few *crucial points* to observe:
- *_everything_ *is in*wide chars*, i.e. wchar_t's:
    - the command line to be used in CreateProcess is an arrow of wide chars
    - sprintf is s_*w*_printf - using wide chars
    - the swprintf's format (a wide string by its _*L*_ prefix) specifies %_*l*_s (instead of the usual %s) to denote wide chars strings
    - accordingly, the strings sent to swprintf (i.e. ParamStr()) are themselves Unicode strings, as is the called executable (_*L*_"Uncde.exe")
    - STARTUPINFO_*W*_ is used instead of the classic STARTUPINFO
    - and of course, CreateProcess_*W*_ is called instead of the classic CreateProcess.

Executing that function simply starts another copy of the executable, passing it - in Unicode! - the parameters it has itself received. That new copy can do the same... and so on - yielding this cascade:


So, all in all, the solution to this thread's subject is simple:
- stop musing about codepages and simply *switch to Unicode/wide char* (wchar_t) everywhere, including using STARTUPINFO_*W*_ and CreateProcess_*W*_ in spawning the user's executable
- naturally, the user themselves need to be able to receive and correctly handle the Unicode parameters
- which seems to suggest that using just a different @SUBJECT@ will not make it (since you can't send /just one/ of the parameters in Unicode and the rest in Ansi) - a different "Run program" will probably be needed in FiltaQuilla to maintain backward compatbility, say Run Unicode.

**Comment by DebuggedTBonce** [comment]
Ok, I see that the embedded images didn't make it through into the the read (which is a shame). So here goes at least fot the all-important function to call, in text:

``` 
void __fastcall TForm1::SpawnButClick(TObject *Sender)
{
   STARTUPINFOW strtinfo;
   PROCESS_INFORMATION retinfo;
   wchar_t cmdline[1024];

   swprintf(cmdline,
            L"%ls \"%ls\" \"%ls\" \"%ls\" \"%ls\" \"%ls\" \"%ls\" \"%ls\" \"%ls\"",
            L"Uncde.exe",
            ParamStr(1),ParamStr(2),ParamStr(3),ParamStr(4),
            ParamStr(5),ParamStr(6),ParamStr(7),ParamStr(8));
   memset(&strtinfo,0,sizeof(STARTUPINFOW));       // Set all fields to default.
   strtinfo.cb=sizeof(STARTUPINFOW);                    // That one must be set.
   CreateProcessW(NULL,cmdline,NULL,NULL,0,0,NULL,NULL,&strtinfo,&retinfo);
}
```

I'll upload the images once to my main machine.

**Comment by RealRaven2000** [comment]
I took the liberty of reformating your last 2 posts on github (mainly removing unnecessary line breaks) - but unfortunately if you use email replies you cannot do markdown, so here is the proper code example marked as C++ code:

```c++
void __fastcall TForm1::SpawnButClick(TObject *Sender)
{
   STARTUPINFOW strtinfo;
   PROCESS_INFORMATION retinfo;
   wchar_t cmdline[1024];

   swprintf(cmdline,
            L"%ls \"%ls\" \"%ls\" \"%ls\" \"%ls\" \"%ls\" \"%ls\" \"%ls\" \"%ls\"",
            L"Uncde.exe",
            ParamStr(1),ParamStr(2),ParamStr(3),ParamStr(4),
            ParamStr(5),ParamStr(6),ParamStr(7),ParamStr(8));
   memset(&strtinfo,0,sizeof(STARTUPINFOW));       // Set all fields to default.
   strtinfo.cb=sizeof(STARTUPINFOW);                    // That one must be set.
   CreateProcessW(NULL,cmdline,NULL,NULL,0,0,NULL,NULL,&strtinfo,&retinfo);
}
```

Bear in mind that not everything clever you do in C is also possible in JavaScript  as JavaScript is a weakly typed language.

**Comment by DebuggedTBonce** [comment]
Hi,

Back to my regular machine now...

Your formatting of my code is absolutely correct, thank you. However, here's a better version, devoid of anything Embarcadero (meaning should compile in any C/C++ compiler) and safer:

```
#define MAXSPAWNCMDLEN 1024

bool spawnexeW(wchar_t *exepath, wchar_t **params)
{
  STARTUPINFOW strtinfo;
  PROCESS_INFORMATION retinfo;
  wchar_t cmdline[MAXSPAWNCMDLEN], *p;
  unsigned i;

  p=_wcspcpy(cmdline,L"\"");
  p=_wcspcpy(p,exepath);
  p=_wcspcpy(p,L"\"");
  for(i=0;i<8 && params[i];i++)
  {
    if((p-cmdline)+wcslen(params[i])+3>=MAXSPAWNCMDLEN)    // >= : terminal \0.
      break;                                         // Would overflow cmdline.
    p=_wcspcpy(p,L" \"");
    p=_wcspcpy(p,params[i]);
    p=_wcspcpy(p,L"\"");
  }
  memset(&strtinfo,0,sizeof(STARTUPINFOW));       // Set all fields to default.
  strtinfo.cb=sizeof(STARTUPINFOW);                    // That one must be set.
  return CreateProcessW(NULL,cmdline,NULL,NULL,0,0,NULL,NULL,&strtinfo,&retinfo);
}
```

Also, here are the screenshots that didn't make it through via email previously:

An application that shows parameters passed to it in Unicode (a.k.a. wide-chars):

![image](https://user-images.githubusercontent.com/92902505/157576464-dea9fc01-9b4f-4b0d-8f4b-eba1a3129617.png)

...where all sorts of "exotic" alphabets are used (all of those snippets are from TB, btw). I'd be willing to share the application for your tests, if you're interested.

The parameters in that shot were specified by the IDE, but they can be passed along by the application to subsequent copies of itself, by calling the _spawnexeW_ function above:

![image](https://user-images.githubusercontent.com/92902505/157576870-f0f7507c-3c38-4daf-a512-97f28f033877.png)


Now to your latest post. I had no idea you'd write it in JavaScript! Argh, those untyped languages... Anyway, I'd see a few possibilities:

- Can you call C (possibly in a DLL) from JavaScript? Python can, so I guess it should be possible. If so, then I'd be willing to create a DLL (32 and/or 64 bits) with the above _spawnexeW_ , so that you could call it to start a process with Unicode parameters
- If not, then I guess one possibility would be to encode the Unicode (which you _do_ get from TB) in Hex and send it (in UTF-8) instead of the original Unicode. Naturally, that would be different parameters, say @SUBJECTHEX16@ etc., and it would be up to the receiving application to translate the Hex into wide-char strings on reception.

Hope that may help, and thanks for your effort!

DebuggedTBonce

**Comment by RealRaven2000** [comment]
> Now to your latest post. I had no idea you'd write it in JavaScript! Argh, those untyped languages... Anyway, I'd see a few possibilities:
> 
> * Can you call C (possibly in a DLL) from JavaScript? Python can, so I guess it should be possible. If so, then I'd be willing to create a DLL (32 and/or 64 bits) with the above _spawnexeW_ , so that you could call it to start a process with Unicode parameters

no - web extensions are not allowed calling binary components. With my Add-on I have some lee-way because even though I am restricted to JavaSCript I can currently still call system functions which are available to the Thunderbird front-end (hence the dire warning about full access to the PC when you install FiltaQuilla). The other thing is that when legacy extension packaged DLLs there was a separate (much slower, delays of weeks are possible) review process - on top of that you are bound to specific OS so I generally do not want to support binaries and keep it "purely scripted".

> * If not, then I guess one possibility would be to encode the Unicode (which you _do_ get from TB) in Hex and send it (in UTF-8) instead of the original Unicode. Naturally, that would be different parameters, say @SUBJECTHEX16@ etc., and it would be up to the receiving application to translate the Hex into wide-char strings on reception.

I think I may have already tried UTF-8 as format, but my command line testing utility (still in Windows 7 64bit) may not be the best way of checking. New parameters such as @SUBJECTHEX16@ may be ok, but maybe a global switch (or selection dropdown for different character formats) in FQ settings dialog might be an easier option for both end users and me.

**Comment by DebuggedTBonce** [comment]
Ok, I may understand your unwillingness to call a DLL. So that, I believe, leaves only my second option.

There is clearly a confusion regarding the UTF-8 encoding: I see you tried to call a converter... but that will **definitely not** produce a conversion into hexadecimals! It will just try to extract whatever 8-bit _corresponding_ characters it can from the 16-bit Unicode you receive from TB - but it won't get anywhere on e.g. "Другие способы познакомиться с Wrike" (apart from "Wrike", as indeed reported by Null-Kelvin on Aug 27, 2021).

No, what you need is to **convert the Unicode** you get from TB (and I can see in your post of Oct 29, 2021 that _you do_ receive it OK) **to hexadecimals before** giving it to Ci.nsIProcess,run( ). And actually, a simple Google search reveals that **the appropriate function exists in JavaScript(!)**: https://stackoverflow.com/questions/21647928/javascript-unicode-string-to-hex .

So I think it's settled: to produce @SUBJECTHEX16@ and @AUTHORHEX16@, you simply send Ci.nsIProcess,run( ) the hex strings obtained with the **hexEncode()** function given in the Stack Overflow.

Thank you,

DebuggedTBonce

**Comment by DebuggedTBonce** [comment]
SOLVED!!!

And pretty trivially: simply call nsIProcess' run**w**() instead of run(): the terminal **w** will make it accept parameters as **w**ide-chars (i.e. Unicode) - exactly as we have always wished.
Dramatic advantage: you just need to append that **w** - I'd guess it will send the Unicode with _**no other change** to your existing code_.

Would you be so kind to try? Thanks!

DebuggedTBonce

P.S. Solution found here: https://stackoverflow.com/questions/27087289/firefox-add-on-execute-shell-command-parameters-encoding . Literally: "To get javascript engine correctly pass parameters in unicode encoding [_as they are_ in TB] it is necessary to use method runw of the nsIProcess. In such case arguments, using UTF-16, will be passed to the executable on its command line." (!)
The guy even explicitly mentions the Russian alphabet - I guess @Null-Kelvin will appreciate. ;-)

**Comment by DebuggedTBonce** [comment]
One last thing: I'm of course willing to share with you the small program that supplied my screenshots above, for your testing (as it shows what **unicode** parameters were received. Or alternatively, if you'd prefer to send me a "beta" of FQ with the run**w**() [see my previous post], I'd be glad to test it and report.

**Comment by RealRaven2000** [comment]
> SOLVED!!!
> 
> And pretty trivially: simply call nsIProcess' run**w**() instead of run(): the terminal **w** will make it accept parameters as **w**ide-chars (i.e. Unicode) - exactly as we have always wished. Dramatic advantage: you just need to append that **w** - I'd guess it will send the Unicode with _**no other change** to your existing code_.

I found it in the archive (the xpcom documentation is not hosted on Mozilla's mdn anymore, as they are discouraging Add-on authors to use it - as you know xpcom is not exposed to web extensions / new mail extensions):

https://udn.realityripple.com/docs/Mozilla/Tech/XPCOM/Reference/Interface/nsIProcess#runw()

the question is, if it is backwards compatible enough could it just be a global switch in FiltaQuilla or should it be a separate action ("Select a Program to run [wide character set]") -  the main problem with the later solution is that it a very long entry for the actions dropdown, and is very likely to be truncated. So I would prefer the former solution a simple checkbox "use wide characters".

**Comment by RealRaven2000** [comment]
I made a test version, but "ordinary" parameters all have to be UTF-16 encoded to work? In that case I would use the following function to convert any "naturally UTF8 encoded parts" to UTF16:

```javascript
/**
 * Convert a UTF8 string to UTF16.
 * @param {String} input
 * @returns {String}
 */
utf8To16: function (input) {
    var _escape = function(s) {
        function q(c) {
            c = c.charCodeAt();
            return '%' + (c<16 ? '0' : '') + c.toString(16).toUpperCase();
        }
        return s.replace(/[\x00-),:-?[-^`{-\xFF]/g, q);
    };
    try{
        return decodeURIComponent(_escape(input));
    }catch (URIError) {
        //include invalid character, cannot convert
        return input;
    }
},
```

I use a Windows Command Script "FQTest.cmd" which simply appends the parameters to a text file:

```cmd
@ECHO OFF
cd /d %~dp0
chcp 65001 >NUL 2>NUL
ECHO %1 > Output.txt
ECHO %2 >> Output.txt
```

![image](https://user-images.githubusercontent.com/6314674/157864504-bcf8e23d-f583-4a61-a07c-b84ad824f336.png)

to append the subject of a message "[FiltaQuilla] issue 102 - Другие способы познакомиться с Wrike[FiltaQuilla]"

the result:

"[FiltaQuilla] issue 102 - @C385 A?>A>1K ?>7=0:><8BLAO A Wrike" 
3ed8dde2-0685-d10f-acfe-ac702c294f05@gmail.com 

so obviously I need to convert the strings to UTF-16 as well. Still strange that some words came through. I will try converting all parameters through the conversion routine above next...

**Comment by RealRaven2000** [comment]
Here is a test version - I managed to integrate the utf16 conversion routine successfully:

[filtaquilla-3.4pre65.zip](https://github.com/RealRaven2000/FiltaQuilla/files/8231761/filtaquilla-3.4pre65.zip)

To install, download the zip file and drop it (without unpacking) into Thunderbird Add-ons Manager. With Filtaquilla it is also important to restart Thunderbird if you want to avoid duplication of its Action dropdowns. (WE are still not able to remove custom actions).

To enable new encoding behavior, simply check the new option "_Transmit Parameters as UTF-16_":

![image](https://user-images.githubusercontent.com/6314674/157866392-f05ec7db-9bdf-455f-8bec-1ad74fdc39fe.png)

**Comment by DebuggedTBonce** [comment]
Hi,

Thanks for the "beta" - received and installed OK:
![image](https://user-images.githubusercontent.com/92902505/157909861-dc99c920-0dfa-4d79-a92e-093c00af6cc6.png)

I will now create a new filter to call the app that produced my screenshots above with @SUBJECT@ etc., to test the reception of UTF-16. Will report shortly.

Best,

DebuggedTBonce

**Comment by DebuggedTBonce** [comment]
**YES, it works!!** BUT...

I confirm that sending the Unicode strings works as intended. Here's proof, where I sent @SUBJECT@,@DATE@ to my test application:

![image](https://user-images.githubusercontent.com/92902505/157931140-66fd8afb-aa3a-45cc-bd24-9434597c3cb6.png)

-- I'm sure @Null-Kelvin will be pleased (as am I!).  :-)

**BUT**, that being said, I think that it's a bad idea to have that checkbox in FQ preferences - because it **applies to all filters**. And that's a serious problem regarding backward compatibility: people may have programs they run that are unable to swallow UTF-16 as parameters, yet may well wish to _also_ run other programs (on selected messages) that _require_ UTF-16. With that check box applying to all filters, they cannot have both.

I think that keeping Run Program with UTF-8 parameters and adding a **Run Program UTF-16** as a new functionality to specify in a filter (in place of the current Run Program) would solve that dilemma.
But of course, that's just a humble suggestion - you are the FQ designer.

Can't wait to have the UTF-16 parameters for real, thank you!
Best,

DebuggedTBonce

**Comment by RealRaven2000** [comment]
> I think that keeping Run Program with UTF-8 parameters and adding a **Run Program UTF-16** as a new functionality to specify in a filter (in place of the current Run Program) would solve that dilemma.
> But of course, that's just a humble suggestion - you are the FQ designer.

I think people wanting both might be an exception, but for them we could add a @UTF16@ parameter? It would switch the original Run Program to the new encoding without checking the box. I bet the majority will find the additional checkbox sufficient and just decide for one flavor only.

**Comment by DebuggedTBonce** [comment]
Ok, I understand the additional complexity of adding a new filter _functionality_ (a new entry in the functionality drop-list, checking the syntax, etc.). Requiring **@UTF16@** as the first parameter for those who did **not** check the UTF-16 checkbox in the Options looks like a reasonable compromise (although you will have to explain in the Help that it will _not_ be actually sent to the called program).
I was just thinking of other users, given that myself I'm fine with the checkbox.

Just a very last suggestion if I may: you may wish to call the checkbox "**Run in UTF-16**", instead of the current "Transmit Parameters as UTF-16" - it's a far shorter text and you already have a pretty limited space in the Options window.

**Comment by RealRaven2000** [comment]
I changed the configuration behavior; so instead of hardcoding it to "Only use UTF-16 or UTF-8 exclusively" through a checkbox, this is now merely used as the "default value". And you can switch to the other behavior on a per case basis by adding the parameter `@UTF16@` or [@UTF8@](url) anywhere in the comma separated list:

![image](https://user-images.githubusercontent.com/6314674/158021703-87d9c127-0e78-4f23-90b9-088f74461176.png)

I renamed the option accordingly and provided a tooltip that shows up as soon as you click the new `?` button.

![image](https://user-images.githubusercontent.com/6314674/158021767-3622bca0-7a58-4151-a92e-1c8ef24d4bdf.png)

Test version: [filtaquilla-3.4pre69.zip](https://github.com/RealRaven2000/FiltaQuilla/files/8237758/filtaquilla-3.4pre69.zip)

To install, download the zip file and drop it (without unpacking) into Thunderbird Add-ons Manager. Probably restart Thunderbird.

**Comment by RealRaven2000** [comment]
Implemented in 3.4, released 18/03/2022
